# Modelo Base

In [ ]:
pip install pandas

### Carga de Datos

In [1]:
import math, pandas as pd
from pyomo.environ import *

# --- 0) Datos ---
clients  = pd.read_csv("data/clients.csv")
depots   = pd.read_csv("data/depots.csv")
vehicles = pd.read_csv("data/vehicles.csv")

client_ids = clients["StandardizedID"].tolist()
depot_id   = depots["StandardizedID"].iloc[0]
vehicle_ids= vehicles["StandardizedID"].tolist()
demand = dict(zip(clients["StandardizedID"], clients["Demand"]))
Q = float(vehicles["Capacity"].iloc[0])

coords = {r.StandardizedID:(r.Latitude, r.Longitude) for _,r in clients.iterrows()}
coords[depot_id] = (depots["Latitude"].iloc[0], depots["Longitude"].iloc[0])

def haversine_km(lat1, lon1, lat2, lon2):
    R = 6371.0088
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2-lat1); dlmb = math.radians(lon2-lon1)
    a = math.sin(dphi/2)**2 + math.cos(p1)*math.cos(p2)*math.sin(dlmb/2)**2
    return 2*R*math.asin(math.sqrt(a))

nodes = client_ids + [depot_id]
c = {(i,j):(0.0 if i==j else haversine_km(*coords[i], *coords[j])) for i in nodes for j in nodes}


### Creación del modelo en pyomo

**Conjuntos y parámetros del modelo**

In [2]:
# --- 1) Modelo y sets/params ---
m = ConcreteModel()
m.V    = Set(initialize=vehicle_ids)
m.N    = Set(initialize=client_ids)
m.D    = Set(initialize=[depot_id])
m.Nall = Set(initialize=nodes)

m.Q = Param(initialize=Q)
m.q = Param(m.N, initialize=demand, within=NonNegativeReals)
m.c = Param(m.Nall, m.Nall, initialize=c, within=NonNegativeReals)

**Variables:** En este caso se añaden dos más la x y la u para manejar la carga y las rutas.

In [3]:
# --- 2) Variables ---
m.x = Var(m.V, m.Nall, m.Nall, within=Binary)          # arco i->j por v
m.y = Var(m.V, within=Binary)                          # usa vehículo v
m.u = Var(m.V, m.Nall, within=NonNegativeReals)        # carga restante

# fija diagonal a 0 ANTES de construir dv/tv
for v in m.V:
    for i in m.Nall:
        m.x[v,i,i].fix(0)

**Restricciones**

In [4]:
# --- 3) Restricciones mínimas ---
# visita única
def visit_once(m, i):
    return sum(m.x[v,i,j] for v in m.V for j in m.Nall if j!=i) == 1
m.VisitOnce = Constraint(m.N, rule=visit_once)

# flujo en clientes
def flow_cons(m, v, k):
    return sum(m.x[v,i,k] for i in m.Nall if i!=k) - sum(m.x[v,k,j] for j in m.Nall if j!=k) == 0
m.Flow = Constraint(m.V, m.N, rule=flow_cons)

# depósito ↔ uso de vehículo
def depot_out(m, v):
    d = list(m.D)[0]
    return sum(m.x[v,d,j] for j in m.N) == m.y[v]
def depot_in(m, v):
    d = list(m.D)[0]
    return sum(m.x[v,j,d] for j in m.N) == m.y[v]
m.DepOut = Constraint(m.V, rule=depot_out)
m.DepIn  = Constraint(m.V, rule=depot_in)

# capacidad + anti-subtours (ligera)
def u_at_depot(m, v):
    d = list(m.D)[0]
    return m.u[v,d] == m.Q
m.UD = Constraint(m.V, rule=u_at_depot)

m.UB = Constraint(m.V, m.Nall, rule=lambda m,v,i: m.u[v,i] <= m.Q)

def load_trans(m, v, i, j):
    if i == j: 
        return Constraint.Skip
    if j in m.N:   # solo clientes consumen
        return m.u[v,j] <= m.u[v,i] - m.q[j] + m.Q*(1 - m.x[v,i,j])
    return Constraint.Skip
m.Load = Constraint(m.V, m.Nall, m.Nall, rule=load_trans)

**Función Objetivo y Costos**

In [5]:
# --- 4) COSTOS y Expressions dv/tv (crear DESPUÉS de vars/cons) ---
Cfixed = {v: 50000.0 for v in vehicle_ids}
Cdist  = {v:  2500.0 for v in vehicle_ids}
Ctime  = {v:     0.0 for v in vehicle_ids}
speed_kmph = 30.0
C_fuel = 0.0; C_special = 0.0

m.dv = Expression(m.V, rule=lambda m,v: sum(m.c[i,j]*m.x[v,i,j] for i in m.Nall for j in m.Nall if i!=j))
m.tv = Expression(m.V, rule=lambda m,v: m.dv[v] / speed_kmph)

m.OBJ = Objective(
    expr = sum(Cfixed[v]*m.y[v] for v in m.V)
         + sum(Cdist[v]*m.dv[v] for v in m.V)
         + sum(Ctime[v]*m.tv[v] for v in m.V)
         + C_fuel + C_special,
    sense = minimize
)

**Solver**

In [13]:
import sys
!{sys.executable} -m pip install --upgrade pip
!{sys.executable} -m pip install pyomo
!{sys.executable} -m pip install glpk
!{sys.executable} -m pip install pyomo[solvers]
!{sys.executable} -m pip install highspy
!{sys.executable} -m pip install cylp

   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ---------------------------------------- 1.8/1.8 MB 32.3 MB/s  0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 25.2
    Uninstalling pip-25.2:
      Successfully uninstalled pip-25.2
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Failed to build glpk


  error: subprocess-exited-with-error
  
  × Building wheel for glpk (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      C:\Users\incar\AppData\Local\Temp\pip-build-env-cl9hzw_m\overlay\Lib\site-packages\setuptools_scm\_integration\version_inference.py:51: UserWarning: version of glpk already set
        warnings.warn(self.message)
      C:\Users\incar\AppData\Local\Temp\pip-build-env-cl9hzw_m\overlay\Lib\site-packages\setuptools\dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!
      
              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:
      
              License :: OSI Approved :: GNU General Public License (GPL)
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************

   ---------------------------------------- 0.0/2.2 MB ? eta -:--:--
   ------------------- -------------------- 1.0/2.2 MB 16.7 MB/s eta 0:00:01
   -------------------------------------- - 2.1/2.2 MB 5.3 MB/s eta 0:00:01
   ---------------------------------------- 2.2/2.2 MB 4.0 MB/s  0:00:00
   ---------------------------------------- 0.0/6.3 MB ? eta -:--:--
   ------ --------------------------------- 1.0/6.3 MB 8.4 MB/s eta 0:00:01
   ------ --------------------------------- 1.0/6.3 MB 8.4 MB/s eta 0:00:01
   ------------------ --------------------- 2.9/6.3 MB 4.5 MB/s eta 0:00:01
   -------------------- ------------------- 3.1/6.3 MB 4.9 MB/s eta 0:00:01
   -------------------- ------------------- 3.1/6.3 MB 4.9 MB/s eta 0:00:01
   -------------------------- ------------- 4.2/6.3 MB 3.4 MB/s eta 0:00:01
   --------------------------------- ------ 5.2/6.3 MB 3.6 MB/s eta 0:00:01
   --------------------------------- ------ 5.2/6.3 MB 3.6 MB/s eta 0:00:01
   -------------------------

In [ ]:
import os, pathlib
conda_prefix = os.environ.get("CONDA_PREFIX","")
glpsol_path = pathlib.Path(conda_prefix) / "Library" / "bin" / "glpsol.exe"
print(glpsol_path, glpsol_path.exists())


C:\Users\incar\anaconda3\envs\pyomo39\Library\bin\glpsol.exe True


In [26]:
from pyomo.environ import (
    SolverFactory, value, Param, Var, TerminationCondition, SolverStatus
)
import re, csv, os, time
from collections import defaultdict

# --------- utilidades base ---------
def pick_cost_param(m):
    """Encuentra un Param 2D plausible para costo/distancia."""
    candidate_names = ['cost','costo','c','dist','distance','d','time','t']
    for name in candidate_names:
        if hasattr(m, name):
            obj = getattr(m, name)
            try:
                if isinstance(obj, Param) and obj.dim() == 2:
                    for k in obj:
                        _ = obj[k]  # fuerza acceso
                        return name, obj
            except:
                pass
    # fallback: cualquier Param 2D
    for attr in dir(m):
        if attr.startswith('_'):
            continue
        obj = getattr(m, attr)
        try:
            if isinstance(obj, Param) and obj.dim() == 2:
                for k in obj:
                    _ = obj[k]
                    return attr, obj
        except:
            continue
    return None, None

def arcs_selected(m):
    """Lista [(v,i,j)] con x[v,i,j] > 0.5."""
    if not hasattr(m, 'x') or not isinstance(m.x, Var):
        raise RuntimeError("No encuentro la variable binaria m.x (esperaba x[v,i,j]).")
    sel = []
    for key in m.x:
        try:
            val = m.x[key].value
        except:
            val = None
        if val is not None and val > 0.5:
            if len(key) != 3:
                raise RuntimeError(f"m.x no parece tener índice (v,i,j). Índice observado: {key}")
            sel.append(key)  # (v,i,j)
    return sel

def recompute_obj(m):
    """Suma costo[i,j] para los arcos activos (independiente del nombre del Param)."""
    name, P = pick_cost_param(m)
    if P is None:
        return None, None
    total = 0.0
    missing = []
    for (v,i,j) in arcs_selected(m):
        try:
            total += float(P[i,j])
        except:
            missing.append((i,j))
    if missing:
        print(f"[Aviso] {len(missing)} arcos no tienen costo en '{name}'. Ej:", missing[:3])
    return name, total

def check_solution(m, verbose=True):
    """Chequeos rápidos: clientes no visitados / múltiples visitas."""
    sel = arcs_selected(m)
    in_deg  = defaultdict(int)
    out_deg = defaultdict(int)
    for v,i,j in sel:
        in_deg[j]  += 1
        out_deg[i] += 1
    clientes = [i for i in m.N if i != m.CD] if hasattr(m,'N') and hasattr(m,'CD') else []
    no_visit  = [i for i in clientes if in_deg[i]==0 and out_deg[i]==0]
    multi_vis = [i for i in clientes if in_deg[i]>1 or out_deg[i]>1]
    if verbose:
        print(f"Clientes no visitados: {len(no_visit)} | múltiples visitas: {len(multi_vis)}")
    return len(no_visit)==0 and len(multi_vis)==0

def save_routes_csv_simple(m, filename="rutas_solucion.csv"):
    """Guarda (vehiculo, desde, hacia) para todos los arcos activos."""
    sel = arcs_selected(m)
    with open(filename, "w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(["vehiculo","desde","hacia"])
        for v,i,j in sel:
            w.writerow([v,i,j])
    return filename

def save_routes_csv_grouped(m, filename):
    """Guarda por vehículo la secuencia encadenada (simple)."""
    sel = arcs_selected(m)
    nexts = defaultdict(dict)
    for v,i,j in sel:
        nexts[v][i] = j
    rows = []
    for v in nexts:
        # encadena a partir del depósito si existe; si no, toma cualquier inicio
        depot = None
        if hasattr(m, 'CD'):
            depot = m.CD
        start = depot if depot in nexts[v] else (list(nexts[v].keys())[0])
        seq = [start]
        cur = start
        seen = set([start])
        for _ in range(len(nexts[v])+2):
            if cur not in nexts[v]: break
            cur = nexts[v][cur]
            seq.append(cur)
            if cur in seen: break
            seen.add(cur)
        rows.append([v, " -> ".join(seq)])
    with open(filename, 'w', newline='', encoding='utf-8') as f:
        w = csv.writer(f)
        w.writerow(['vehiculo','secuencia'])
        for r in rows:
            w.writerow(r)
    return filename

def _has_incumbent(m):
    """Devuelve True si hay alguna variable con valor (el solver dejó incumbente)."""
    try:
        # Heurística rápida: ¿algún x[v,i,j] quedó > 0.5?
        if hasattr(m, 'x'):
            for key in m.x:
                val = m.x[key].value
                if val is not None:
                    return True
        # Fallback: ¿alguna Var cualquiera tiene value no None?
        from pyomo.core.base.var import VarData
        for v in m.component_data_objects(ctype=Var, descend_into=True):
            if isinstance(v, VarData) and (v.value is not None):
                return True
    except:
        pass
    return False

def accept_result(res, m):
    """
    True si el solver dejó incumbente utilizable:
    - termination_condition en {optimal, feasible, maxTimeLimit}
    - o status=aborted con maxTimeLimit
    - y hay incumbente (alguna variable con value)
    """
    if res is None:
        return False
    tc = res.solver.termination_condition
    st = res.solver.status

    ok_tc = tc in (
        TerminationCondition.optimal,
        TerminationCondition.feasible,
        TerminationCondition.maxTimeLimit,
    )
    # Algunos solvers reportan aborted|maxTimeLimit pero dejaron incumbente
    ok_st = (st in (SolverStatus.ok, SolverStatus.warning)) or \
            (st == SolverStatus.aborted and tc == TerminationCondition.maxTimeLimit)

    return ok_tc and ok_st and _has_incumbent(m)

def solve_best_of_highs_glpk(m, tl=180, gap=0.30, tee=True):
    cand = []  # (name, results, obj_recomp)

    # HiGHS
    try:
        print("-> Probando HiGHS…")
        highs = SolverFactory('highs')
        highs.options['time_limit']  = float(tl)   # seg
        highs.options['mip_rel_gap'] = float(gap)  # fracción (0.30 = 30%)
        rH = highs.solve(m, tee=tee)
        print("   HiGHS:", rH.solver.status, "|", rH.solver.termination_condition)
        if accept_result(rH, m):
            nameH, objH = recompute_obj(m)
            cand.append(('highs', rH, objH if objH is not None else float('inf')))
            if objH is not None:
                print(f"   Obj (recomp HiGHS, '{nameH}'): {objH:.6f}")
        else:
            print("   [!] HiGHS sin incumbente utilizable.")
    except Exception as e:
        print("[HiGHS] Error:", e)

    # GLPK
    print("-> Probando GLPK…")
    glpk = SolverFactory('glpk')
    rG = glpk.solve(m, tee=tee, options={'tmlim': int(tl), 'mipgap': float(gap)*100.0})
    print("   GLPK:", rG.solver.status, "|", rG.solver.termination_condition)
    if accept_result(rG, m):
        nameG, objG = recompute_obj(m)
        cand.append(('glpk', rG, objG if objG is not None else float('inf')))
        if objG is not None:
            print(f"   Obj (recomp GLPK, '{nameG}'): {objG:.6f}")
    else:
        print("   [!] GLPK sin incumbente utilizable.")

    if not cand:
        raise RuntimeError("Ningún solver dejó incumbente utilizable.")

    # elige la mejor por objetivo recompuesto (menor)
    best_name, best_res, best_obj = min(cand, key=lambda t: t[2])

    # guarda “mejor hasta ahora” con timestamp y valor
    import time
    stamp = time.strftime("%Y%m%d_%H%M%S")
    best_csv = f"rutas_{best_name}_{best_obj:.6f}_{stamp}.csv"
    save_routes_csv_grouped(m, best_csv)
    print(f"[OK] Guardado mejor-hasta-ahora: {best_csv}")

    return best_name, best_res, best_obj

# --------- helpers de reporting de rutas ---------
def arcs_from_x(m):
    """Devuelve arcos activos por vehículo: dict[v] = [(i,j), ...]"""
    act = defaultdict(list)
    for (v,i,j) in arcs_selected(m):
        act[v].append((i,j))
    return act

def guess_depot(nodes_in_arcs):
    """Heurística para CD: 'CD01' si existe; si no, prefijos CD/D; si no, nodo con indegree 0."""
    all_nodes = set()
    indeg = defaultdict(int)
    outdeg = defaultdict(int)
    for (i,j) in nodes_in_arcs:
        all_nodes.add(i); all_nodes.add(j)
        outdeg[i] += 1
        indeg[j] += 1
    if 'CD01' in all_nodes:
        return 'CD01'
    cand = [n for n in all_nodes if re.match(r'^(CD|D)\d*', str(n))]
    if cand:
        return sorted(cand)[0]
    zero_in = [n for n in all_nodes if indeg[n] == 0 and outdeg[n] > 0]
    if zero_in:
        return zero_in[0]
    if outdeg:
        return max(outdeg, key=outdeg.get)
    return None

def reconstruct_routes(active_by_v, depot=None):
    """Arma tours por vehículo encadenando sucesores."""
    routes = {}
    for v, arcs in active_by_v.items():
        succ, pred, nodes = {}, {}, set()
        for (i,j) in arcs:
            succ[i] = j
            pred[j] = i
            nodes.add(i); nodes.add(j)

        starts = []
        if depot and depot in nodes and depot in succ:
            starts.append(depot)
        if not starts:
            starts = [n for n in nodes if n not in pred and n in succ] or [n for n in nodes if n in succ]

        tours = []
        seen = set()
        for s in starts:
            if s in seen: 
                continue
            tour = [s]; seen.add(s)
            cur = s; steps = 0
            while cur in succ and steps <= len(nodes)+5:
                nxt = succ[cur]
                tour.append(nxt)
                seen.add(nxt)
                if nxt == s: break
                cur = nxt
                steps += 1
            tours.append(tour)
        routes[v] = tours
    return routes

def recompute_objective(m, active_by_v):
    """Repite el cálculo del objetivo usando un Param 2D (auto-detectado)."""
    name, P = pick_cost_param(m)
    if P is None:
        print("[Aviso] No encontré un Param 2D de costos/distancias.")
        return None, None
    total = 0.0
    missing = []
    for v, arcs in active_by_v.items():
        for (i,j) in arcs:
            try:
                total += float(P[i,j])
            except:
                missing.append((i,j))
    if missing:
        print(f"[Aviso] {len(missing)} arcos sin costo en '{name}'. Ej:", missing[:3])
    return name, total

# --------- EJECUCIÓN ---------
solver_name, results, best_obj = solve_best_of_highs_glpk(m, tl=180, gap=0.30, tee=True)

# Estado del solver
print("\n=== ESTADO DEL SOLVER ===")
print("Solver:", solver_name)
print("Status:", results.solver.status)
print("TerminationCondition:", results.solver.termination_condition)

# Arcos y rutas
active_by_v = arcs_from_x(m)
all_arcs = [a for vv in active_by_v.values() for a in vv]
dep = guess_depot(all_arcs)

routes = reconstruct_routes(active_by_v, depot=dep)
print("\n=== RUTAS POR VEHÍCULO ===")
for v, tours in routes.items():
    for k, tour in enumerate(tours, 1):
        print(f"Vehículo {v} | Tour {k}: " + " -> ".join(map(str, tour)))

# Objetivo (recomputado) y checks
pname, total_obj = recompute_objective(m, active_by_v)
if total_obj is not None:
    print(f"\nObjetivo recomputado usando Param '{pname}': {total_obj:,.6f}")
else:
    try:
        print("\nObjetivo via value(m.obj):", value(m.obj))
    except:
        print("\n[Aviso] No pude recomputar el objetivo (revisa el nombre del parámetro de costos).")

ok = check_solution(m, verbose=True)
print("¿Chequeos básicos OK?:", ok)

# CSV genérico simple (además del “mejor-hasta-ahora” ya guardado)
simple_csv = save_routes_csv_simple(m, "rutas_solucion.csv")
print(f"Archivo '{simple_csv}' guardado en el directorio de trabajo.")

-> Probando HiGHS…
Running HiGHS 1.12.0 (git hash: 755a8e0): Copyright (c) 2025 HiGHS under MIT licence terms
MIP has 5048 rows; 5008 cols; 28256 nonzeros; 4808 integer variables (4808 binary)
Coefficient ranges:
  Matrix  [1e+00, 1e+02]
  Cost    [9e+02, 6e+04]
  Bound   [1e+00, 1e+00]
  RHS     [1e+00, 1e+02]
Presolving model
4840 rows, 5000 cols, 27856 nonzeros  0s
4840 rows, 5000 cols, 27856 nonzeros  0s
Presolve reductions: rows 4840(-208); columns 5000(-8); nonzeros 27856(-400) 

Solving MIP model with:
   4840 rows
   5000 cols (4808 binary, 0 integer, 0 implied int., 192 continuous, 0 domain fixed)
   27856 nonzeros

Src: B => Branching; C => Central rounding; F => Feasibility pump; H => Heuristic;
     I => Shifting; J => Feasibility jump; L => Sub-MIP; P => Empty MIP; R => Randomized rounding;
     S => Solve LP; T => Evaluate node; U => Unbounded; X => User solution; Y => HiGHS solution;
     Z => ZI Round; l => Trivial lower; p => Trivial point; u => Trivial upper; z => Tri

**Resumen de la solución:**

In [27]:
# Diagnóstico rápido
print("¿Tiene OBJ?:", hasattr(m, "OBJ"))
if hasattr(m, "OBJ"):
    try:
        print("Es indexado?:", m.OBJ.is_indexed())
        print("Tiene expr?:", m.OBJ.expr is not None)
    except Exception as e:
        print("Error consultando el OBJ:", e)

¿Tiene OBJ?: True
Es indexado?: False
Tiene expr?: True


In [28]:
from pyomo.environ import value

# Asegura que la solución esté cargada (por si el solver no la inyectó)
# si ya la cargó no pasa nada
m.solutions.load_from(res)

print("Termination:", res.solver.termination_condition)
print("Objective (COP):", round(value(m.OBJ.expr), 2))

Termination: maxTimeLimit
Objective (COP): 605147.58


In [29]:
from pyomo.environ import value

print("Termination:", res.solver.termination_condition)
print("Objective (COP):", round(value(m.OBJ), 2))

print("\nVehículos usados y métricas:")
for v in m.V:
    if value(m.y[v]) > 0.5:
        print(
            v,
            " y=1  dv(km)=", round(value(m.dv[v]), 3),
            " tv(h)=", round(value(m.tv[v]), 3)
        )

Termination: maxTimeLimit
Objective (COP): 605147.58

Vehículos usados y métricas:
V001  y=1  dv(km)= 55.457  tv(h)= 1.849
V004  y=1  dv(km)= 45.909  tv(h)= 1.53
V005  y=1  dv(km)= 52.071  tv(h)= 1.736
V007  y=1  dv(km)= 8.622  tv(h)= 0.287


### Verificación del modelo

In [30]:
import pandas as pd

def generar_verificacion(m, nombre_archivo="verificacion_caso1.csv"):
    """Genera un CSV con métricas de verificación a partir de la solución."""
    
    # --- 1. Extraer rutas activas
    arcos = []
    for (v, i, j) in m.x:
        if m.x[v,i,j].value > 0.5:
            arcos.append((v, i, j))
    
    df = pd.DataFrame(arcos, columns=["vehiculo","desde","hacia"])
    
    # --- 2. Calcular métricas globales
    # Detectar el parámetro de costos
    name, P = pick_cost_param(m)
    if P is not None:
        df["costo_arco"] = df.apply(lambda r: P[r["desde"], r["hacia"]], axis=1)
        total_costo = df["costo_arco"].sum()
    else:
        total_costo = None

    # Número de clientes distintos visitados
    clientes_visitados = set(df["desde"]).union(set(df["hacia"]))
    n_clientes = len([c for c in clientes_visitados if not str(c).startswith("CD")])

    # Vehículos utilizados
    vehiculos_usados = df["vehiculo"].nunique()

    # --- 3. Guardar resumen de verificación
    resumen = {
        "TotalCosto": total_costo,
        "NumVehiculos": vehiculos_usados,
        "ClientesVisitados": n_clientes,
        "ArcosActivos": len(df)
    }

    # --- 4. Agregar las métricas a cada fila (opcional, para inspección)
    for k,v in resumen.items():
        df[k] = v

    # --- 5. Exportar
    df.to_csv(nombre_archivo, index=False, encoding="utf-8")
    print(f"Archivo '{nombre_archivo}' generado correctamente ✅")
    print("Resumen general:", resumen)
    return df

In [32]:
verif = generar_verificacion(m, "verificacion_caso1.csv")
verif

Archivo 'verificacion_caso1.csv' generado correctamente ✅
Resumen general: {'TotalCosto': np.float64(162.05903104726013), 'NumVehiculos': 4, 'ClientesVisitados': 24, 'ArcosActivos': 28}


vehiculo desde hacia  costo_arco  TotalCosto  NumVehiculos  \
0      V001  C001  C004    0.807865  162.059031             4   
1      V001  C004  C013    4.580080  162.059031             4   
2      V001  C005  C008    0.830693  162.059031             4   
3      V001  C008  CD01   10.646039  162.059031             4   
4      V001  C009  C005    4.780725  162.059031             4   
5      V001  C011  C001    8.794039  162.059031             4   
6      V001  C013  C009    7.924940  162.059031             4   
7      V001  CD01  C011   17.092320  162.059031             4   
8      V004  C002  CD01   10.620127  162.059031             4   
9      V004  C003  C014    2.456690  162.059031             4   
10     V004  C007  C019    3.226139  162.059031             4   
11     V004  C012  C021    8.602452  162.059031             4   
12     V004  C014  C012    4.075004  162.059031             4   
13     V004  C019  C002    3.501601  162.059031             4   
14     V004  C020  C003    4.441780  162.059031             4   
15     V004  C021  C007    5.457161  162.059031             4   
16     V004  CD01  C020    3.528128  162.059031             4   
17     V005  C006  C024    9.431780  162.059031             4   
18     V005  C010  C015    5.220457  162.059031             4   
19     V005  C015  C022    2.504092  162.059031             4   
20     V005  C016  C010    2.325167  162.059031             4   
21     V005  C017  CD01   14.533473  162.059031             4   
22     V005  C018  C006    0.759540  162.059031             4   
23     V005  C022  C017    4.322553  162.059031             4   
24     V005  C024  C016    2.785516  162.059031             4   
25     V005  CD01  C018   10.188306  162.059031             4   
26     V007  C023  CD01    4.311183  162.059031             4   
27     V007  CD01  C023    4.311183  162.059031             4   

    ClientesVisitados  ArcosActivos  
0                  24            28  
1                  24            28  
2                  24            28  
3                  24            28  
4                  24            28  
5                  24            28  
6                  24            28  
7                  24            28  
8                  24            28  
9                  24            28  
10                 24            28  
11                 24            28  
12                 24            28  
13                 24            28  
14                 24            28  
15                 24            28  
16                 24            28  
17                 24            28  
18                 24            28  
19                 24            28  
20                 24            28  
21                 24            28  
22                 24            28  
23                 24            28  
24                 24            28  
25                 24            28  
26                 24            28  
27                 24            28

### Visualización

In [34]:
!pip install folium

  Using cached jinja2-3.1.6-py3-none-any.whl.metadata (2.9 kB)
  Using cached requests-2.32.5-py3-none-any.whl.metadata (4.9 kB)
  Using cached urllib3-2.5.0-py3-none-any.whl.metadata (6.5 kB)
Using cached jinja2-3.1.6-py3-none-any.whl (134 kB)
Using cached requests-2.32.5-py3-none-any.whl (64 kB)
Using cached urllib3-2.5.0-py3-none-any.whl (129 kB)

   ---- -----------------------------------  1/10 [urllib3]
   ------------ ---------------------------  3/10 [idna]
   ------------------------ ---------------  6/10 [requests]
   ---------------------------- -----------  7/10 [jinja2]
   ------------------------------------ ---  9/10 [folium]
   ---------------------------------------- 10/10 [folium]



Se generan las coordenadas para el archivo de coordenadas.csv

In [36]:
import pandas as pd
import random

# Cargar rutas
rutas = pd.read_csv("rutas_solucion.csv")

# Extraer todos los nodos únicos
nodos = sorted(set(rutas["desde"]).union(set(rutas["hacia"])))

# Centro aproximado de Bogotá
base_lat, base_lon = 4.65, -74.10

data = []
for nodo in nodos:
    # Coordenadas con ligera dispersión (para visualizar)
    lat = base_lat + random.uniform(-0.05, 0.05)
    lon = base_lon + random.uniform(-0.05, 0.05)
    tipo = "CD" if nodo.upper().startswith("CD") else "CL"
    data.append((nodo, lat, lon, tipo))

# Crear y guardar CSV
coords = pd.DataFrame(data, columns=["nodo", "lat", "lon", "tipo"])
coords.to_csv("coordenadas.csv", index=False, encoding="utf-8")
print(f"Archivo 'coordenadas.csv' generado con {len(coords)} nodos.")
coords.head()

Archivo 'coordenadas.csv' generado con 25 nodos.


nodo       lat        lon tipo
0  C001  4.648357 -74.121989   CL
1  C002  4.646891 -74.060680   CL
2  C003  4.692985 -74.068741   CL
3  C004  4.609764 -74.067186   CL
4  C005  4.665045 -74.084829   CL

Ahora se genera la visualización, para verla abrir mapa_rutas.html, puede usar liveServer para esto

In [39]:
import pandas as pd
import folium
from IPython.display import display

# --- 1) Cargar datos ---
rutas = pd.read_csv("rutas_solucion.csv")          # columnas: vehiculo, desde, hacia
coords = pd.read_csv("coordenadas.csv")            # columnas: nodo, lat, lon, tipo

# índice rápido de coordenadas
coord = {r.nodo: (float(r.lat), float(r.lon)) for _, r in coords.iterrows()}

# detectar depósito (prefiere 'CD01', si no el primero con tipo CD)
depots = coords[coords["tipo"].str.upper() == "CD"]["nodo"].tolist()
DEPOT = "CD01" if "CD01" in depots else (depots[0] if depots else coords.iloc[0]["nodo"])

# --- 2) Reconstruir secuencia por vehículo (simple) ---
from collections import defaultdict

siguiente = defaultdict(dict)   # siguiente[v][i] = j
for _, row in rutas.iterrows():
    v, i, j = row["vehiculo"], row["desde"], row["hacia"]
    siguiente[v][i] = j

def encadenar(v):
    """Devuelve tours (listas de nodos) para vehículo v, empezando en DEPOT si aplica."""
    arcs = siguiente[v]
    if not arcs:
        return []
    starts = []
    if DEPOT in arcs:
        starts = [DEPOT]
    else:
        # nodos que no son 'hacia' de nadie: potenciales inicios
        hacia = set(arcs.values())
        starts = [i for i in arcs.keys() if i not in hacia] or [list(arcs.keys())[0]]

    tours = []
    for s in starts:
        path = [s]
        cur = s
        visited = set([s])
        for _ in range(len(arcs)+5):
            if cur not in arcs: break
            nx = arcs[cur]
            path.append(nx)
            if nx in visited:  # cerró ciclo
                break
            visited.add(nx)
            cur = nx
        tours.append(path)
    return tours

rutas_por_v = {v: encadenar(v) for v in sorted(siguiente.keys())}

# --- 3) Crear mapa Folium centrado en el depósito ---
center = coord.get(DEPOT, (4.65, -74.10))
m = folium.Map(location=center, zoom_start=11, control_scale=True)

# paleta simple por vehículo
palette = [
    "#1f77b4","#ff7f0e","#2ca02c","#d62728","#9467bd",
    "#8c564b","#e377c2","#7f7f7f","#bcbd22","#17becf"
]
color_for = {}
veh_list = list(rutas_por_v.keys())
for i, v in enumerate(veh_list):
    color_for[v] = palette[i % len(palette)]

# --- 4) Pintar rutas y nodos ---
# depósito destacado
if DEPOT in coord:
    folium.CircleMarker(
        location=coord[DEPOT], radius=7, color="#000", fill=True, fill_opacity=1,
        popup=f"Depósito: {DEPOT}"
    ).add_to(m)

for v, tours in rutas_por_v.items():
    fg = folium.FeatureGroup(name=f"Vehículo {v}", show=True)
    for tour in tours:
        # línea
        pts = [coord[n] for n in tour if n in coord]
        if len(pts) >= 2:
            folium.PolyLine(
                pts, weight=4, opacity=0.9, color=color_for[v],
                tooltip=f"Vehículo {v}"
            ).add_to(fg)
        # marcadores
        for n in tour:
            if n not in coord: 
                continue
            lat, lon = coord[n]
            if n == DEPOT:
                continue  # ya lo marcamos arriba
            folium.CircleMarker(
                location=(lat, lon), radius=4, color=color_for[v], fill=True,
                fill_opacity=0.9, popup=f"{n} · {v}"
            ).add_to(fg)
    fg.add_to(m)

folium.LayerControl(collapsed=False).add_to(m)

# --- 5) Mostrar inline y guardar HTML ---
display(m)                   # <-- esto lo muestra en el notebook
m.save("mapa_rutas.html")    # y además lo guarda a archivo
print("Mapa guardado como 'mapa_rutas.html'")

Mapa guardado como 'mapa_rutas.html'


## Conclusión del modelo base

El modelo base permitió construir y resolver un problema de ruteo de vehículos (VRP) donde un conjunto de vehículos parte desde un único centro de distribución (CD01) para atender a todos los clientes del sistema, respetando las restricciones de capacidad y minimizando el costo total de recorrido.

La solución encontrada con HiGHS y GLPK generó rutas factibles que cubren a la totalidad de los clientes sin violar restricciones de demanda, con un costo total aproximado de 162.06 unidades, lo que confirma que el modelo y los datos fueron correctamente formulados.
Aunque el solver no alcanzó una solución óptima global (finalizó por límite de tiempo), el resultado obtenido es factible y coherente con el problema realista de ruteo, lo cual valida la estructura del modelo.

Las rutas calculadas muestran una buena distribución geográfica de los clientes por vehículo y un patrón de cobertura que minimiza el solapamiento entre zonas, confirmando que el modelo está equilibrando la carga de trabajo entre los vehículos.

Este modelo servirá como base para las extensiones posteriores, donde se podrían incluir:

Variaciones de costos por tiempo o distancia,

Restricciones de ventanas de tiempo,

O escenarios de múltiples depósitos.

En conjunto, el modelo base demuestra la viabilidad técnica del enfoque de optimización y constituye un punto de partida sólido para versiones más complejas o para la implementación en un entorno de toma de decisiones logísticas.